{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# **Human Protein CDS**: Extract Codons Usage\n",
    "\n",
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: num2words in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (0.5.12)\n",
      "Requirement already satisfied: docopt>=0.6.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from num2words) (0.6.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install num2words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from num2words import num2words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize Codons data\n",
    "\n",
    "CODONS_LIST = [\n",
    "    \"TTT\",\n",
    "    \"TTC\",\n",
    "    \"TTA\",\n",
    "    \"TTG\",\n",
    "    \"TCT\",\n",
    "    \"TCC\",\n",
    "    \"TCA\",\n",
    "    \"TCG\",\n",
    "    \"TAT\",\n",
    "    \"TAC\",\n",
    "    \"TAA\",\n",
    "    \"TAG\",\n",
    "    \"TGT\",\n",
    "    \"TGC\",\n",
    "    \"TGA\",\n",
    "    \"TGG\",\n",
    "    \"CTT\",\n",
    "    \"CTC\",\n",
    "    \"CTA\",\n",
    "    \"CTG\",\n",
    "    \"CCT\",\n",
    "    \"CCC\",\n",
    "    \"CCA\",\n",
    "    \"CCG\",\n",
    "    \"CAT\",\n",
    "    \"CAC\",\n",
    "    \"CAA\",\n",
    "    \"CAG\",\n",
    "    \"CGT\",\n",
    "    \"CGC\",\n",
    "    \"CGA\",\n",
    "    \"CGG\",\n",
    "    \"ATT\",\n",
    "    \"ATC\",\n",
    "    \"ATA\",\n",
    "    \"ATG\",\n",
    "    \"ACT\",\n",
    "    \"ACC\",\n",
    "    \"ACA\",\n",
    "    \"ACG\",\n",
    "    \"AAT\",\n",
    "    \"AAC\",\n",
    "    \"AAA\",\n",
    "    \"AAG\",\n",
    "    \"AGT\",\n",
    "    \"AGC\",\n",
    "    \"AGA\",\n",
    "    \"AGG\",\n",
    "    \"GTT\",\n",
    "    \"GTC\",\n",
    "    \"GTA\",\n",
    "    \"GTG\",\n",
    "    \"GCT\",\n",
    "    \"GCC\",\n",
    "    \"GCA\",\n",
    "    \"GCG\",\n",
    "    \"GAT\",\n",
    "    \"GAC\",\n",
    "    \"GAA\",\n",
    "    \"GAG\",\n",
    "    \"GGT\",\n",
    "    \"GGC\",\n",
    "    \"GGA\",\n",
    "    \"GGG\",\n",
    "]\n",
    "\n",
    "CODONS_AMINO_ACIDS_MAP = {\n",
    "    \"TTT\": \"[F] Phenylalanine\",\n",
    "    \"TTC\": \"[F] Phenylalanine\",\n",
    "    \"TTA\": \"[L] Leucine\",\n",
    "    \"TTG\": \"[L] Leucine\",\n",
    "    \"TCT\": \"[S] Serine\",\n",
    "    \"TCC\": \"[S] Serine\",\n",
    "    \"TCA\": \"[S] Serine\",\n",
    "    \"TCG\": \"[S] Serine\",\n",
    "    \"TAT\": \"[Y] Tyrosine\",\n",
    "    \"TAC\": \"[Y] Tyrosine\",\n",
    "    \"TAA\": \"[*] STOP\",\n",
    "    \"TAG\": \"[*] STOP\",\n",
    "    \"TGT\": \"[C] Cysteine\",\n",
    "    \"TGC\": \"[C] Cysteine\",\n",
    "    \"TGA\": \"[*] STOP\",\n",
    "    \"TGG\": \"[W] Tryptophan\",\n",
    "    \"CTT\": \"[L] Leucine\",\n",
    "    \"CTC\": \"[L] Leucine\",\n",
    "    \"CTA\": \"[L] Leucine\",\n",
    "    \"CTG\": \"[L] Leucine\",\n",
    "    \"CCT\": \"[P] Proline\",\n",
    "    \"CCC\": \"[P] Proline\",\n",
    "    \"CCA\": \"[P] Proline\",\n",
    "    \"CCG\": \"[P] Proline\",\n",
    "    \"CAT\": \"[H] Histidine\",\n",
    "    \"CAC\": \"[H] Histidine\",\n",
    "    \"CAA\": \"[Q] Glutamine\",\n",
    "    \"CAG\": \"[Q] Glutamine\",\n",
    "    \"CGT\": \"[R] Arginine\",\n",
    "    \"CGC\": \"[R] Arginine\",\n",
    "    \"CGA\": \"[R] Arginine\",\n",
    "    \"CGG\": \"[R] Arginine\",\n",
    "    \"ATT\": \"[I] Isoleucine\",\n",
    "    \"ATC\": \"[I] Isoleucine\",\n",
    "    \"ATA\": \"[I] Isoleucine\",\n",
    "    \"ATG\": \"[M] Methionine\",\n",
    "    \"ACT\": \"[T] Threonine\",\n",
    "    \"ACC\": \"[T] Threonine\",\n",
    "    \"ACA\": \"[T] Threonine\",\n",
    "    \"ACG\": \"[T] Threonine\",\n",
    "    \"AAT\": \"[N] Asparagine\",\n",
    "    \"AAC\": \"[N] Asparagine\",\n",
    "    \"AAA\": \"[K] Lysine\",\n",
    "    \"AAG\": \"[K] Lysine\",\n",
    "    \"AGT\": \"[S] Serine\",\n",
    "    \"AGC\": \"[S] Serine\",\n",
    "    \"AGA\": \"[R] Arginine\",\n",
    "    \"AGG\": \"[R] Arginine\",\n",
    "    \"GTT\": \"[V] Valine\",\n",
    "    \"GTC\": \"[V] Valine\",\n",
    "    \"GTA\": \"[V] Valine\",\n",
    "    \"GTG\": \"[V] Valine\",\n",
    "    \"GCT\": \"[A] Alanine\",\n",
    "    \"GCC\": \"[A] Alanine\",\n",
    "    \"GCA\": \"[A] Alanine\",\n",
    "    \"GCG\": \"[A] Alanine\",\n",
    "    \"GAT\": \"[D] Aspartic Acid\",\n",
    "    \"GAC\": \"[D] Aspartic Acid\",\n",
    "    \"GAA\": \"[E] Glutamic Acid\",\n",
    "    \"GAG\": \"[E] Glutamic Acid\",\n",
    "    \"GGT\": \"[G] Glycine\",\n",
    "    \"GGC\": \"[G] Glycine\",\n",
    "    \"GGA\": \"[G] Glycine\",\n",
    "    \"GGG\": \"[G] Glycine\",\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Method that counts codons in given CDS string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "#\n",
    "# \n",
    "def count_codons(cds):\n",
    "    # print(type(cds))\n",
    "    counts = [0] * 64\n",
    "    codons = [cds[i:i+3] for i in range(0, len(cds), 3)]\n",
    "    # print(codons)\n",
    "    codons_count = dict(Counter(codons))\n",
    "    \n",
    "    array_count = list(map(lambda cod: 0 if not cod in codons_count else codons_count[cod], CODONS_LIST))\n",
    "    return array_count"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genes information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#\n",
    "# Uncomment to re-create the tsv file <<<<<<<<<<<\n",
    "#\n",
    "genes_and_proteins_tsv = open(\"./human_genes_and_proteins.tsv\", \"w\")\n",
    "proteins = []\n",
    "\n",
    "with open(\"./cds_from_genomic.fna\") as handle:\n",
    "    for record in SeqIO.parse(handle, \"fasta\"):\n",
    "        # print(record.id)\n",
    "        desc = record.description\n",
    "        tokens = [ \"id=\" + record.id.split(\"|\")[1] ]\n",
    "        tokens = tokens + [protein for protein in [i.split(\"]\")[0] for i in desc.split(\"[\")[1:]] if protein.split(\"=\")[0] in [\"gene\", \"protein\", \"protein_id\"]]\n",
    "        tokens.append(\"cds=\" + str(record.seq))\n",
    "        proteins.append(tokens)\n",
    "\n",
    "coding_proteins = [protein for protein in proteins if len(protein) == 5]\n",
    "just_gene_protein_id = [[value.split(\"=\")[1] for value in rec] for rec in coding_proteins]\n",
    "# print(coding_proteins)\n",
    "# print(just_gene_protein_id)\n",
    "\n",
    "genes_and_proteins_tsv.write(\"ncbi_id\\tgene\\tprotein_id\\tprotein_name\\t\" + \"\\t\".join(CODONS_LIST) + \"\\n\")\n",
    "# print(\"ncbi_id\\tgene\\tprotein_id\\tprotein_name\\t\" + \"\\t\".join(CODONS_LIST) + \"\\n\")\n",
    "for p in just_gene_protein_id:\n",
    "    genes_and_proteins_tsv.write(p[0] + \"\\t\" + p[1] + \"\\t\" + p[3] + \"\\t\" + p[2] + \"\\t\" + \"\\t\".join(map(str, count_codons(p[4]))) + \"\\n\")\n",
    "    # print(p[0] + \"\\t\" + p[2] + \"/\" + p[1] + \"\\t\" + \"\\t\".join(map(str, count_codons(p[3]))))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>TTT</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TCC</th>\n",
       "      <th>TCA</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TAT</th>\n",
       "      <th>...</th>\n",
       "      <th>GCA</th>\n",
       "      <th>GCG</th>\n",
       "      <th>GAT</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GGC</th>\n",
       "      <th>GGA</th>\n",
       "      <th>GGG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20029 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         protein  TTT  TTC  TTA  TTG  TCT  TCC  TCA  TCG  TAT  ...  GCA  GCG  \\\n",
       "gene                                                           ...             \n",
       "A1BG           1    1    1    1    1    1    1    1    1    1  ...    1    1   \n",
       "A1CF          14   14   14   14   14   14   14   14   14   14  ...   14   14   \n",
       "A2M            5    5    5    5    5    5    5    5    5    5  ...    5    5   \n",
       "A2ML1          7    7    7    7    7    7    7    7    7    7  ...    7    7   \n",
       "A3GALT2        1    1    1    1    1    1    1    1    1    1  ...    1    1   \n",
       "...          ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   \n",
       "ZYG11A         9    9    9    9    9    9    9    9    9    9  ...    9    9   \n",
       "ZYG11B         3    3    3    3    3    3    3    3    3    3  ...    3    3   \n",
       "ZYX            8    8    8    8    8    8    8    8    8    8  ...    8    8   \n",
       "ZZEF1          8    8    8    8    8    8    8    8    8    8  ...    8    8   \n",
       "ZZZ3          19   19   19   19   19   19   19   19   19   19  ...   19   19   \n",
       "\n",
       "         GAT  GAC  GAA  GAG  GGT  GGC  GGA  GGG  \n",
       "gene                                             \n",
       "A1BG       1    1    1    1    1    1    1    1  \n",
       "A1CF      14   14   14   14   14   14   14   14  \n",
       "A2M        5    5    5    5    5    5    5    5  \n",
       "A2ML1      7    7    7    7    7    7    7    7  \n",
       "A3GALT2    1    1    1    1    1    1    1    1  \n",
       "...      ...  ...  ...  ...  ...  ...  ...  ...  \n",
       "ZYG11A     9    9    9    9    9    9    9    9  \n",
       "ZYG11B     3    3    3    3    3    3    3    3  \n",
       "ZYX        8    8    8    8    8    8    8    8  \n",
       "ZZEF1      8    8    8    8    8    8    8    8  \n",
       "ZZZ3      19   19   19   19   19   19   19   19  \n",
       "\n",
       "[20029 rows x 65 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "gpdf = pd.read_csv(\"./human_genes_and_proteins.tsv\", sep=\"\\t\")\n",
    "# print(gpdf)\n",
    "\n",
    "gpdf.groupby([\"gene\"]).count()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python code to process protein Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_count = gpdf[gpdf.columns[2:]].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTT    1611118\n",
      "TTC    1627492\n",
      "TTA     833315\n",
      "TTG    1278242\n",
      "TCT    1607659\n",
      "        ...   \n",
      "GAG    3771832\n",
      "GGT    1025541\n",
      "GGC    1867973\n",
      "GGA    1616254\n",
      "GGG    1442185\n",
      "Length: 64, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(final_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAG 30731\n",
      "GAG 3771832\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(final_count.idxmin(), final_count[final_count.idxmin()])\n",
    "print(final_count.idxmax(), final_count[final_count.idxmax()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAG      30731\n",
      "TAA      38263\n",
      "TGA      69014\n",
      "TCG     385999\n",
      "CGT     431421\n",
      "        ...   \n",
      "AAG    3016615\n",
      "GAA    3189350\n",
      "CAG    3384175\n",
      "CTG    3397427\n",
      "GAG    3771832\n",
      "Length: 64, dtype: int64\n",
      "94510990\n"
     ]
    }
   ],
   "source": [
    "# print(final_count.idxmin())\n",
    "# print(final_count[final_count.idxmin()])\n",
    "\n",
    "# print(final_count.idxmax())\n",
    "# print(final_count[final_count.idxmax()])\n",
    "\n",
    "print(final_count.sort_values())\n",
    "\n",
    "print(final_count.sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ninety-four million, five hundred and ten thousand, nine hundred and ninety Codons in Human Proteins\n",
      "two hundred and eighty-three million, five hundred and thirty-two thousand, nine hundred and seventy Nucleotides in Human Proteins\n"
     ]
    }
   ],
   "source": [
    "print(num2words(94510990) + \" Codons in Human Proteins\")\n",
    "print(num2words(94510990 * 3) + \" Nucleotides in Human Proteins\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(gpdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "448e3f0ae365410c91c918c6183b3155",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "import nglview as nv\n",
    "\n",
    "nv.show_file(\"./sample.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
