{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# **Viral Genomes**: Extract Codons Usage\n",
    "\n",
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports and Globals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: num2words in /Users/ishikanelaturu/Library/jupyterlab-desktop/jlab_server/lib/python3.8/site-packages (0.5.12)\n",
      "Requirement already satisfied: docopt>=0.6.2 in /Users/ishikanelaturu/Library/jupyterlab-desktop/jlab_server/lib/python3.8/site-packages (from num2words) (0.6.2)\n"
     ]
    }
   ],
   "source": [
    "!pip3 install num2words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from num2words import num2words"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize Codons data\n",
    "\n",
    "CODONS_LIST = [\n",
    "    \"TTT\",\n",
    "    \"TTC\",\n",
    "    \"TTA\",\n",
    "    \"TTG\",\n",
    "    \"TCT\",\n",
    "    \"TCC\",\n",
    "    \"TCA\",\n",
    "    \"TCG\",\n",
    "    \"TAT\",\n",
    "    \"TAC\",\n",
    "    \"TAA\",\n",
    "    \"TAG\",\n",
    "    \"TGT\",\n",
    "    \"TGC\",\n",
    "    \"TGA\",\n",
    "    \"TGG\",\n",
    "    \"CTT\",\n",
    "    \"CTC\",\n",
    "    \"CTA\",\n",
    "    \"CTG\",\n",
    "    \"CCT\",\n",
    "    \"CCC\",\n",
    "    \"CCA\",\n",
    "    \"CCG\",\n",
    "    \"CAT\",\n",
    "    \"CAC\",\n",
    "    \"CAA\",\n",
    "    \"CAG\",\n",
    "    \"CGT\",\n",
    "    \"CGC\",\n",
    "    \"CGA\",\n",
    "    \"CGG\",\n",
    "    \"ATT\",\n",
    "    \"ATC\",\n",
    "    \"ATA\",\n",
    "    \"ATG\",\n",
    "    \"ACT\",\n",
    "    \"ACC\",\n",
    "    \"ACA\",\n",
    "    \"ACG\",\n",
    "    \"AAT\",\n",
    "    \"AAC\",\n",
    "    \"AAA\",\n",
    "    \"AAG\",\n",
    "    \"AGT\",\n",
    "    \"AGC\",\n",
    "    \"AGA\",\n",
    "    \"AGG\",\n",
    "    \"GTT\",\n",
    "    \"GTC\",\n",
    "    \"GTA\",\n",
    "    \"GTG\",\n",
    "    \"GCT\",\n",
    "    \"GCC\",\n",
    "    \"GCA\",\n",
    "    \"GCG\",\n",
    "    \"GAT\",\n",
    "    \"GAC\",\n",
    "    \"GAA\",\n",
    "    \"GAG\",\n",
    "    \"GGT\",\n",
    "    \"GGC\",\n",
    "    \"GGA\",\n",
    "    \"GGG\",\n",
    "]\n",
    "\n",
    "CODONS_AMINO_ACIDS_MAP = {\"TTT\": \"[F] Phenylalanine\",\n",
    "    \"TTC\": \"[F] Phenylalanine\",\n",
    "    \"TTA\": \"[L] Leucine\",\n",
    "    \"TTG\": \"[L] Leucine\",\n",
    "    \"TCT\": \"[S] Serine\",\n",
    "    \"TCC\": \"[S] Serine\",\n",
    "    \"TCA\": \"[S] Serine\",\n",
    "    \"TCG\": \"[S] Serine\",\n",
    "    \"TAT\": \"[Y] Tyrosine\",\n",
    "    \"TAC\": \"[Y] Tyrosine\",\n",
    "    \"TAA\": \"[*] STOP\",\n",
    "    \"TAG\": \"[*] STOP\",\n",
    "    \"TGT\": \"[C] Cysteine\",\n",
    "    \"TGC\": \"[C] Cysteine\",\n",
    "    \"TGA\": \"[*] STOP\",\n",
    "    \"TGG\": \"[W] Tryptophan\",\n",
    "    \"CTT\": \"[L] Leucine\",\n",
    "    \"CTC\": \"[L] Leucine\",\n",
    "    \"CTA\": \"[L] Leucine\",\n",
    "    \"CTG\": \"[L] Leucine\",\n",
    "    \"CCT\": \"[P] Proline\",\n",
    "    \"CCC\": \"[P] Proline\",\n",
    "    \"CCA\": \"[P] Proline\",\n",
    "    \"CCG\": \"[P] Proline\",\n",
    "    \"CAT\": \"[H] Histidine\",\n",
    "    \"CAC\": \"[H] Histidine\",\n",
    "    \"CAA\": \"[Q] Glutamine\",\n",
    "    \"CAG\": \"[Q] Glutamine\",\n",
    "    \"CGT\": \"[R] Arginine\",\n",
    "    \"CGC\": \"[R] Arginine\",\n",
    "    \"CGA\": \"[R] Arginine\",\n",
    "    \"CGG\": \"[R] Arginine\",\n",
    "    \"ATT\": \"[I] Isoleucine\",\n",
    "    \"ATC\": \"[I] Isoleucine\",\n",
    "    \"ATA\": \"[I] Isoleucine\",\n",
    "    \"ATG\": \"[M] Methionine\",\n",
    "    \"ACT\": \"[T] Threonine\",\n",
    "    \"ACC\": \"[T] Threonine\",\n",
    "    \"ACA\": \"[T] Threonine\",\n",
    "    \"ACG\": \"[T] Threonine\",\n",
    "    \"AAT\": \"[N] Asparagine\",\n",
    "    \"AAC\": \"[N] Asparagine\",\n",
    "    \"AAA\": \"[K] Lysine\",\n",
    "    \"AAG\": \"[K] Lysine\",\n",
    "    \"AGT\": \"[S] Serine\",\n",
    "    \"AGC\": \"[S] Serine\",\n",
    "    \"AGA\": \"[R] Arginine\",\n",
    "    \"AGG\": \"[R] Arginine\",\n",
    "    \"GTT\": \"[V] Valine\",\n",
    "    \"GTC\": \"[V] Valine\",\n",
    "    \"GTA\": \"[V] Valine\",\n",
    "    \"GTG\": \"[V] Valine\",\n",
    "    \"GCT\": \"[A] Alanine\",\n",
    "    \"GCC\": \"[A] Alanine\",\n",
    "    \"GCA\": \"[A] Alanine\",\n",
    "    \"GCG\": \"[A] Alanine\",\n",
    "    \"GAT\": \"[D] Aspartic Acid\",\n",
    "    \"GAC\": \"[D] Aspartic Acid\",\n",
    "    \"GAA\": \"[E] Glutamic Acid\",\n",
    "    \"GAG\": \"[E] Glutamic Acid\",\n",
    "    \"GGT\": \"[G] Glycine\",\n",
    "    \"GGC\": \"[G] Glycine\",\n",
    "    \"GGA\": \"[G] Glycine\",\n",
    "    \"GGG\": \"[G] Glycine\",\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Method that counts codons in given CDS string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "#\n",
    "# \n",
    "def count_codons(cds):\n",
    "    # print(type(cds))\n",
    "    counts = [0] * 64\n",
    "    codons = [cds[i:i+3] for i in range(0, len(cds), 3)]\n",
    "    # print(codons)\n",
    "    codons_count = dict(Counter(codons))\n",
    "    \n",
    "    array_count = list(map(lambda cod: 0 if not cod in codons_count else codons_count[cod], CODONS_LIST))\n",
    "    return array_count"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genes information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "genes_and_proteins_tsv = open(\"./viral_genes_and_proteins.tsv\", \"w\")\n",
    "\n",
    "# Read the GBFF file\n",
    "gbff_file = \"./viral.1.genomic.gbff\"\n",
    "genes_and_proteins_tsv.write(\"ncbi_id\\ttax_url\\torganism\\tstrain\\thost\\tgene\\tprotein_id\\tprotein_name\\t\" + \"\\t\".join(CODONS_LIST) + \"\\n\")\n",
    "\n",
    "records = list(SeqIO.parse(gbff_file, \"genbank\"))\n",
    "for record in records:\n",
    "    # Get the name of the virus\n",
    "    ncbi_id = record.id\n",
    "    organism = record.annotations[\"organism\"]\n",
    "    strain = \"UNKNOWN\"\n",
    "    host = \"UNKNOWN\"\n",
    "    taxonomy = \";\".join(record.annotations[\"taxonomy\"])\n",
    "    tax_url = \"ERROR\"\n",
    "    #if \"Human\" in record_str:\n",
    "    for feature in record.features:\n",
    "        if feature.type == \"source\":\n",
    "            if \"host\" in feature.qualifiers:\n",
    "                host = feature.qualifiers[\"host\"][0]\n",
    "            if \"strain\" in feature.qualifiers:\n",
    "                strain = feature.qualifiers[\"strain\"][0]\n",
    "            # print(host, \":\", record.annotations[\"organism\"], \":\", \"/\".join(record.annotations[\"taxonomy\"]))\n",
    "            if len(record.annotations[\"taxonomy\"]) <= 0:\n",
    "                tax_url = host + \"://\" + \"Viruses;\" + organism + \";\" + strain\n",
    "            else:\n",
    "                tax_url = host + \"://\" + taxonomy + \";\" + organism + \";\" + strain\n",
    "        elif feature.type == \"CDS\": # Check if the feature is a CDS (coding sequence)\n",
    "            gene_name = \"UNKNOWN\"\n",
    "            protein_id = \"UNKNOWN\"\n",
    "            protein_name = \"UNKNOWN\"\n",
    "            if \"gene\" in feature.qualifiers:\n",
    "                gene_name = feature.qualifiers[\"gene\"][0]\n",
    "            if \"protein_id\" in feature.qualifiers:\n",
    "                protein_id = feature.qualifiers[\"protein_id\"][0]\n",
    "            if \"product\" in feature.qualifiers:\n",
    "                protein_name = feature.qualifiers[\"product\"][0]                                              \n",
    "            cds = str(feature.extract(record.seq)) # Extract the protein sequence from the CDS\n",
    "            codons_count = \"\\t\".join(map(str, count_codons(cds)))\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            genes_and_proteins_tsv.write(ncbi_id + \"\\t\" + tax_url + \"\\t\" + organism + \"\\t\" + strain + \"\\t\" + host + \"\\t\" + gene_name + \"\\t\" + protein_id + \"\\t\" + protein_name + \"\\t\" + codons_count + \"\\n\")\n",
    "        \n",
    "genes_and_proteins_tsv.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "gpdf = pd.read_csv(\"./viral_genes_and_proteins.tsv\", sep=\"\\t\")\n",
    "# print(gpdf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbi_id</th>\n",
       "      <th>tax_url</th>\n",
       "      <th>strain</th>\n",
       "      <th>host</th>\n",
       "      <th>gene</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_name</th>\n",
       "      <th>TTT</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTA</th>\n",
       "      <th>...</th>\n",
       "      <th>GCA</th>\n",
       "      <th>GCG</th>\n",
       "      <th>GAT</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GGC</th>\n",
       "      <th>GGA</th>\n",
       "      <th>GGG</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>organism</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>unidentified entomopoxvirus</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sclerotinia sclerotiorum mitovirus 1 HC025</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sclerotinia sclerotiorum mitovirus 2</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sclerotinia sclerotiorum mitovirus 3</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sclerotinia sclerotiorum mitovirus 6</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pandoravirus quercus</th>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>...</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "      <td>1185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orpheovirus IHUMI-LCC2</th>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>...</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "      <td>1199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pandoravirus salinus</th>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>...</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "      <td>1430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>African swine fever virus</th>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>...</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "      <td>1557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pandoravirus inopinatum</th>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>...</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "      <td>1839</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11534 rows × 71 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ncbi_id  tax_url  strain  host  \\\n",
       "organism                                                                     \n",
       "unidentified entomopoxvirus                       1        1       1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 1 HC025        1        1       1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 2              1        1       1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 3              1        1       1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 6              1        1       1     1   \n",
       "...                                             ...      ...     ...   ...   \n",
       "Pandoravirus quercus                           1185     1185    1185  1185   \n",
       "Orpheovirus IHUMI-LCC2                         1199     1199    1199  1199   \n",
       "Pandoravirus salinus                           1430     1430    1430  1430   \n",
       "African swine fever virus                      1557     1557    1557  1557   \n",
       "Pandoravirus inopinatum                        1839     1839    1839  1839   \n",
       "\n",
       "                                            gene  protein_id  protein_name  \\\n",
       "organism                                                                     \n",
       "unidentified entomopoxvirus                    1           1             1   \n",
       "Sclerotinia sclerotiorum mitovirus 1 HC025     1           1             1   \n",
       "Sclerotinia sclerotiorum mitovirus 2           1           1             1   \n",
       "Sclerotinia sclerotiorum mitovirus 3           1           1             1   \n",
       "Sclerotinia sclerotiorum mitovirus 6           1           1             1   \n",
       "...                                          ...         ...           ...   \n",
       "Pandoravirus quercus                        1185        1185          1185   \n",
       "Orpheovirus IHUMI-LCC2                      1199        1199          1199   \n",
       "Pandoravirus salinus                        1430        1430          1430   \n",
       "African swine fever virus                   1557        1557          1557   \n",
       "Pandoravirus inopinatum                     1839        1839          1839   \n",
       "\n",
       "                                             TTT   TTC   TTA  ...   GCA   GCG  \\\n",
       "organism                                                      ...               \n",
       "unidentified entomopoxvirus                    1     1     1  ...     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 1 HC025     1     1     1  ...     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 2           1     1     1  ...     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 3           1     1     1  ...     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 6           1     1     1  ...     1     1   \n",
       "...                                          ...   ...   ...  ...   ...   ...   \n",
       "Pandoravirus quercus                        1185  1185  1185  ...  1185  1185   \n",
       "Orpheovirus IHUMI-LCC2                      1199  1199  1199  ...  1199  1199   \n",
       "Pandoravirus salinus                        1430  1430  1430  ...  1430  1430   \n",
       "African swine fever virus                   1557  1557  1557  ...  1557  1557   \n",
       "Pandoravirus inopinatum                     1839  1839  1839  ...  1839  1839   \n",
       "\n",
       "                                             GAT   GAC   GAA   GAG   GGT  \\\n",
       "organism                                                                   \n",
       "unidentified entomopoxvirus                    1     1     1     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 1 HC025     1     1     1     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 2           1     1     1     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 3           1     1     1     1     1   \n",
       "Sclerotinia sclerotiorum mitovirus 6           1     1     1     1     1   \n",
       "...                                          ...   ...   ...   ...   ...   \n",
       "Pandoravirus quercus                        1185  1185  1185  1185  1185   \n",
       "Orpheovirus IHUMI-LCC2                      1199  1199  1199  1199  1199   \n",
       "Pandoravirus salinus                        1430  1430  1430  1430  1430   \n",
       "African swine fever virus                   1557  1557  1557  1557  1557   \n",
       "Pandoravirus inopinatum                     1839  1839  1839  1839  1839   \n",
       "\n",
       "                                             GGC   GGA   GGG  \n",
       "organism                                                      \n",
       "unidentified entomopoxvirus                    1     1     1  \n",
       "Sclerotinia sclerotiorum mitovirus 1 HC025     1     1     1  \n",
       "Sclerotinia sclerotiorum mitovirus 2           1     1     1  \n",
       "Sclerotinia sclerotiorum mitovirus 3           1     1     1  \n",
       "Sclerotinia sclerotiorum mitovirus 6           1     1     1  \n",
       "...                                          ...   ...   ...  \n",
       "Pandoravirus quercus                        1185  1185  1185  \n",
       "Orpheovirus IHUMI-LCC2                      1199  1199  1199  \n",
       "Pandoravirus salinus                        1430  1430  1430  \n",
       "African swine fever virus                   1557  1557  1557  \n",
       "Pandoravirus inopinatum                     1839  1839  1839  \n",
       "\n",
       "[11534 rows x 71 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gpdf.groupby([\"organism\"]).count().sort_values(by=\"ncbi_id\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python code to process protein Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_count = gpdf[gpdf.columns[2:]].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(final_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(final_count.idxmin(), final_count[final_count.idxmin()])\n",
    "print(final_count.idxmax(), final_count[final_count.idxmax()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(final_count.idxmin())\n",
    "# print(final_count[final_count.idxmin()])\n",
    "\n",
    "# print(final_count.idxmax())\n",
    "# print(final_count[final_count.idxmax()])\n",
    "\n",
    "print(final_count.sort_values())\n",
    "\n",
    "print(final_count.sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(num2words(94510990) + \" Codons in Human Proteins\")\n",
    "print(num2words(94510990 * 3) + \" Nucleotides in Human Proteins\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(gpdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import nglview as nv\n",
    "\n",
    "nv.show_file(\"./sample.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "20807aea3061ca722dd3bc51bf0d4dc32e03bed78a4bf526c837e06f0b847991"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
