{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Viral Immunity through Codon Optimization\n",
    "A Comparative Analysis of Genomic and Viral Data\n",
    "\n",
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Abstract\n",
    "\n",
    "REASON: The importance of my research and algorithm will help  scientists find cures & vaccines to viruses more quickly and efficiently.\n",
    "\n",
    "PROBLEM: As humans continue to develop, experiment, and grow, we as a community will face many hardships along the way. One of these hardships is the spread of new and unknown viruses. The process of finding a cure for these viruses is long and inefficient. As this process occurs, millions could die.\n",
    "\n",
    "METHODOLOGY: By using a simple algorithm, I can go through all my data to find the least used codon in human DNA & the most used codon in virus DNA. \n",
    "\n",
    "RESULT: N/A\n",
    "\n",
    "IMPLICATIONS: Using this data, scientists can find the most used codon in virus DNA & the least used codon in human DNA. If one can somehow reprogram how a ribosome processes mRNA in a way such that viruses can no longer be reproduced in a cell, this would stop the virus itself before it can do much harm. Millions of lives could be saved.\n",
    "\n",
    "The importance of my research and algorithm will help  scientists find cures & vaccines to viruses more quickly and efficiently. As humans continue to develop, experiment, and grow, we as a community will face many hardships along the way. One of these hardships is the spread of new and unknown viruses. The process of finding a cure for these viruses is long and inefficient. As this process occurs, millions could die. By using a simple algorithm, I can go through all my data to find the least used codon in human DNA & the most used codon in virus DNA.(RESULT - N/A). Using this data, scientists can find the most used codon in virus DNA & the least used codon in human DNA. If one can somehow reprogram how a ribosome processes mRNA in a way such that viruses can no longer be reproduced in a cell, this would stop the virus itself before it can do much harm. Millions of lives could be saved.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Creation\n",
    "\n",
    "Explain how data got created\n",
    "\n",
    "[Extracting Codons data from Human Reference CDS](./NCBI-GRCh38.p14_GCF_000001405.40/process_human_codons.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: num2words in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (0.5.12)\n",
      "Requirement already satisfied: docopt>=0.6.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from num2words) (0.6.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install num2words\n",
    "\n",
    "from Bio import SeqIO\n",
    "import Bio.Data.CodonTable\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from num2words import num2words\n",
    "\n",
    "import pprint\n",
    "\n",
    "pp = pprint.PrettyPrinter(indent=4)\n",
    "\n",
    "pd.set_option('display.max_rows', None)\n",
    "# pd.options.display.max_rows = 250\n",
    "# pd.options.display.max_seq_items = 250\n",
    "\n",
    "human_df = pd.read_csv(\"./human_genes_and_proteins.tsv\", sep=\"\\t\")\n",
    "viral_df = pd.read_csv(\"./viral_genes_and_proteins.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Codons to Amino Acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n",
      "{   'AAA': ['AAG'],\n",
      "    'AAC': ['AAT'],\n",
      "    'AAG': ['AAA'],\n",
      "    'AAT': ['AAC'],\n",
      "    'ACA': ['ACT', 'ACC', 'ACG'],\n",
      "    'ACC': ['ACT', 'ACA', 'ACG'],\n",
      "    'ACG': ['ACT', 'ACC', 'ACA'],\n",
      "    'ACT': ['ACC', 'ACA', 'ACG'],\n",
      "    'AGA': ['CGT', 'CGC', 'CGA', 'CGG', 'AGG'],\n",
      "    'AGC': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT'],\n",
      "    'AGG': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA'],\n",
      "    'AGT': ['TCT', 'TCC', 'TCA', 'TCG', 'AGC'],\n",
      "    'ATA': ['ATT', 'ATC'],\n",
      "    'ATC': ['ATT', 'ATA'],\n",
      "    'ATG': [],\n",
      "    'ATT': ['ATC', 'ATA'],\n",
      "    'CAA': ['CAG'],\n",
      "    'CAC': ['CAT'],\n",
      "    'CAG': ['CAA'],\n",
      "    'CAT': ['CAC'],\n",
      "    'CCA': ['CCT', 'CCC', 'CCG'],\n",
      "    'CCC': ['CCT', 'CCA', 'CCG'],\n",
      "    'CCG': ['CCT', 'CCC', 'CCA'],\n",
      "    'CCT': ['CCC', 'CCA', 'CCG'],\n",
      "    'CGA': ['CGT', 'CGC', 'CGG', 'AGA', 'AGG'],\n",
      "    'CGC': ['CGT', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
      "    'CGG': ['CGT', 'CGC', 'CGA', 'AGA', 'AGG'],\n",
      "    'CGT': ['CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
      "    'CTA': ['TTA', 'TTG', 'CTT', 'CTC', 'CTG'],\n",
      "    'CTC': ['TTA', 'TTG', 'CTT', 'CTA', 'CTG'],\n",
      "    'CTG': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA'],\n",
      "    'CTT': ['TTA', 'TTG', 'CTC', 'CTA', 'CTG'],\n",
      "    'GAA': ['GAG'],\n",
      "    'GAC': ['GAT'],\n",
      "    'GAG': ['GAA'],\n",
      "    'GAT': ['GAC'],\n",
      "    'GCA': ['GCT', 'GCC', 'GCG'],\n",
      "    'GCC': ['GCT', 'GCA', 'GCG'],\n",
      "    'GCG': ['GCT', 'GCC', 'GCA'],\n",
      "    'GCT': ['GCC', 'GCA', 'GCG'],\n",
      "    'GGA': ['GGT', 'GGC', 'GGG'],\n",
      "    'GGC': ['GGT', 'GGA', 'GGG'],\n",
      "    'GGG': ['GGT', 'GGC', 'GGA'],\n",
      "    'GGT': ['GGC', 'GGA', 'GGG'],\n",
      "    'GTA': ['GTT', 'GTC', 'GTG'],\n",
      "    'GTC': ['GTT', 'GTA', 'GTG'],\n",
      "    'GTG': ['GTT', 'GTC', 'GTA'],\n",
      "    'GTT': ['GTC', 'GTA', 'GTG'],\n",
      "    'TAA': ['TAG', 'TGA'],\n",
      "    'TAC': ['TAT'],\n",
      "    'TAG': ['TAA', 'TGA'],\n",
      "    'TAT': ['TAC'],\n",
      "    'TCA': ['TCT', 'TCC', 'TCG', 'AGT', 'AGC'],\n",
      "    'TCC': ['TCT', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
      "    'TCG': ['TCT', 'TCC', 'TCA', 'AGT', 'AGC'],\n",
      "    'TCT': ['TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
      "    'TGA': ['TAA', 'TAG'],\n",
      "    'TGC': ['TGT'],\n",
      "    'TGG': [],\n",
      "    'TGT': ['TGC'],\n",
      "    'TTA': ['TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
      "    'TTC': ['TTT'],\n",
      "    'TTG': ['TTA', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
      "    'TTT': ['TTC']}\n"
     ]
    }
   ],
   "source": [
    "CODONS_LIST = [\n",
    "    \"TTT\",\n",
    "    \"TTC\",\n",
    "    \"TTA\",\n",
    "    \"TTG\",\n",
    "    \"TCT\",\n",
    "    \"TCC\",\n",
    "    \"TCA\",\n",
    "    \"TCG\",\n",
    "    \"TAT\",\n",
    "    \"TAC\",\n",
    "    \"TAA\",\n",
    "    \"TAG\",\n",
    "    \"TGT\",\n",
    "    \"TGC\",\n",
    "    \"TGA\",\n",
    "    \"TGG\",\n",
    "    \"CTT\",\n",
    "    \"CTC\",\n",
    "    \"CTA\",\n",
    "    \"CTG\",\n",
    "    \"CCT\",\n",
    "    \"CCC\",\n",
    "    \"CCA\",\n",
    "    \"CCG\",\n",
    "    \"CAT\",\n",
    "    \"CAC\",\n",
    "    \"CAA\",\n",
    "    \"CAG\",\n",
    "    \"CGT\",\n",
    "    \"CGC\",\n",
    "    \"CGA\",\n",
    "    \"CGG\",\n",
    "    \"ATT\",\n",
    "    \"ATC\",\n",
    "    \"ATA\",\n",
    "    \"ATG\",\n",
    "    \"ACT\",\n",
    "    \"ACC\",\n",
    "    \"ACA\",\n",
    "    \"ACG\",\n",
    "    \"AAT\",\n",
    "    \"AAC\",\n",
    "    \"AAA\",\n",
    "    \"AAG\",\n",
    "    \"AGT\",\n",
    "    \"AGC\",\n",
    "    \"AGA\",\n",
    "    \"AGG\",\n",
    "    \"GTT\",\n",
    "    \"GTC\",\n",
    "    \"GTA\",\n",
    "    \"GTG\",\n",
    "    \"GCT\",\n",
    "    \"GCC\",\n",
    "    \"GCA\",\n",
    "    \"GCG\",\n",
    "    \"GAT\",\n",
    "    \"GAC\",\n",
    "    \"GAA\",\n",
    "    \"GAG\",\n",
    "    \"GGT\",\n",
    "    \"GGC\",\n",
    "    \"GGA\",\n",
    "    \"GGG\",\n",
    "]\n",
    "CODONS_AMINO_ACIDS_MAP = {\n",
    "    \"TTT\": \"[F] Phenylalanine\",\n",
    "    \"TTC\": \"[F] Phenylalanine\",\n",
    "    \"TTA\": \"[L] Leucine\",\n",
    "    \"TTG\": \"[L] Leucine\",\n",
    "    \"TCT\": \"[S] Serine\",\n",
    "    \"TCC\": \"[S] Serine\",\n",
    "    \"TCA\": \"[S] Serine\",\n",
    "    \"TCG\": \"[S] Serine\",\n",
    "    \"TAT\": \"[Y] Tyrosine\",\n",
    "    \"TAC\": \"[Y] Tyrosine\",\n",
    "    \"TAA\": \"[*] STOP\",\n",
    "    \"TAG\": \"[*] STOP\",\n",
    "    \"TGT\": \"[C] Cysteine\",\n",
    "    \"TGC\": \"[C] Cysteine\",\n",
    "    \"TGA\": \"[*] STOP\",\n",
    "    \"TGG\": \"[W] Tryptophan\",\n",
    "    \"CTT\": \"[L] Leucine\",\n",
    "    \"CTC\": \"[L] Leucine\",\n",
    "    \"CTA\": \"[L] Leucine\",\n",
    "    \"CTG\": \"[L] Leucine\",\n",
    "    \"CCT\": \"[P] Proline\",\n",
    "    \"CCC\": \"[P] Proline\",\n",
    "    \"CCA\": \"[P] Proline\",\n",
    "    \"CCG\": \"[P] Proline\",\n",
    "    \"CAT\": \"[H] Histidine\",\n",
    "    \"CAC\": \"[H] Histidine\",\n",
    "    \"CAA\": \"[Q] Glutamine\",\n",
    "    \"CAG\": \"[Q] Glutamine\",\n",
    "    \"CGT\": \"[R] Arginine\",\n",
    "    \"CGC\": \"[R] Arginine\",\n",
    "    \"CGA\": \"[R] Arginine\",\n",
    "    \"CGG\": \"[R] Arginine\",\n",
    "    \"ATT\": \"[I] Isoleucine\",\n",
    "    \"ATC\": \"[I] Isoleucine\",\n",
    "    \"ATA\": \"[I] Isoleucine\",\n",
    "    \"ATG\": \"[M] Methionine\",\n",
    "    \"ACT\": \"[T] Threonine\",\n",
    "    \"ACC\": \"[T] Threonine\",\n",
    "    \"ACA\": \"[T] Threonine\",\n",
    "    \"ACG\": \"[T] Threonine\",\n",
    "    \"AAT\": \"[N] Asparagine\",\n",
    "    \"AAC\": \"[N] Asparagine\",\n",
    "    \"AAA\": \"[K] Lysine\",\n",
    "    \"AAG\": \"[K] Lysine\",\n",
    "    \"AGT\": \"[S] Serine\",\n",
    "    \"AGC\": \"[S] Serine\",\n",
    "    \"AGA\": \"[R] Arginine\",\n",
    "    \"AGG\": \"[R] Arginine\",\n",
    "    \"GTT\": \"[V] Valine\",\n",
    "    \"GTC\": \"[V] Valine\",\n",
    "    \"GTA\": \"[V] Valine\",\n",
    "    \"GTG\": \"[V] Valine\",\n",
    "    \"GCT\": \"[A] Alanine\",\n",
    "    \"GCC\": \"[A] Alanine\",\n",
    "    \"GCA\": \"[A] Alanine\",\n",
    "    \"GCG\": \"[A] Alanine\",\n",
    "    \"GAT\": \"[D] Aspartic Acid\",\n",
    "    \"GAC\": \"[D] Aspartic Acid\",\n",
    "    \"GAA\": \"[E] Glutamic Acid\",\n",
    "    \"GAG\": \"[E] Glutamic Acid\",\n",
    "    \"GGT\": \"[G] Glycine\",\n",
    "    \"GGC\": \"[G] Glycine\",\n",
    "    \"GGA\": \"[G] Glycine\",\n",
    "    \"GGG\": \"[G] Glycine\",\n",
    "}\n",
    "\n",
    "AMINO_ACIDS_TO_CODONS_MAP = {}\n",
    "for codon in CODONS_AMINO_ACIDS_MAP:\n",
    "    amino_acid = CODONS_AMINO_ACIDS_MAP[codon]\n",
    "    if not amino_acid in AMINO_ACIDS_TO_CODONS_MAP:\n",
    "        AMINO_ACIDS_TO_CODONS_MAP[amino_acid] = []\n",
    "    AMINO_ACIDS_TO_CODONS_MAP[amino_acid]. append(codon)\n",
    "\n",
    "amino_acids_with_alternate_codons = {}\n",
    "for amino_acid in AMINO_ACIDS_TO_CODONS_MAP:\n",
    "    codons_list = AMINO_ACIDS_TO_CODONS_MAP[amino_acid]\n",
    "    if len(codons_list) > 0: # Change this to 1 if we can replace codon with just one alternate\n",
    "        amino_acids_with_alternate_codons[amino_acid] = codons_list\n",
    "\n",
    "print(len(amino_acids_with_alternate_codons))\n",
    "\n",
    "\n",
    "# Build codons with alternates\n",
    "codon_alternates = {}\n",
    "for clist in amino_acids_with_alternate_codons.values():\n",
    "    for codon in clist:\n",
    "        dup_list = clist[:]\n",
    "        dup_list.remove(codon)\n",
    "        codon_alternates[codon] = dup_list\n",
    "\n",
    "pp.pprint(codon_alternates)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAG      30731\n",
      "TAA      38263\n",
      "TGA      69014\n",
      "TCG     385999\n",
      "CGT     431421\n",
      "ACG     527773\n",
      "GCG     560789\n",
      "CCG     593810\n",
      "CGA     605853\n",
      "CTA     705153\n",
      "GTA     725910\n",
      "ATA     760858\n",
      "CGC     825087\n",
      "TTA     833315\n",
      "TGT     979680\n",
      "TGC    1010784\n",
      "GGT    1025541\n",
      "CGG    1031183\n",
      "TGG    1079433\n",
      "CAT    1118497\n",
      "GTT    1119006\n",
      "AGG    1136933\n",
      "TAT    1138776\n",
      "AGA    1249060\n",
      "TAC    1258885\n",
      "GTC    1260452\n",
      "TTG    1278242\n",
      "CAA    1332898\n",
      "CTT    1339874\n",
      "AGT    1340944\n",
      "TCA    1344428\n",
      "ACT    1348416\n",
      "CAC    1379660\n",
      "GGG    1442185\n",
      "ACA    1555392\n",
      "ATT    1556158\n",
      "TCT    1607659\n",
      "TTT    1611118\n",
      "GGA    1616254\n",
      "GCA    1619464\n",
      "TTC    1627492\n",
      "TCC    1636520\n",
      "CTC    1668304\n",
      "ACC    1669290\n",
      "AAC    1722022\n",
      "ATC    1739947\n",
      "AAT    1747788\n",
      "CCA    1800473\n",
      "GCT    1800570\n",
      "CCC    1805566\n",
      "CCT    1833292\n",
      "GGC    1867973\n",
      "AGC    1874608\n",
      "ATG    2027598\n",
      "GAT    2283382\n",
      "GAC    2293356\n",
      "GTG    2432424\n",
      "GCC    2442252\n",
      "AAA    2603836\n",
      "AAG    3016615\n",
      "GAA    3189350\n",
      "CAG    3384175\n",
      "CTG    3397427\n",
      "GAG    3771832\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "human_df.groupby([\"ncbi_id\"]).count()\n",
    "human_codon_count = human_df[human_df.columns[4:]].sum()\n",
    "print(human_codon_count.sort_values())\n",
    "# print(\"------------\")\n",
    "# print(human_codon_count.idxmin(), human_codon_count[human_codon_count.idxmin()])\n",
    "# print(human_codon_count.idxmax(), human_codon_count[human_codon_count.idxmax()])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viral Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAA    5162667\n",
      "AAA    4829550\n",
      "GAT    4540003\n",
      "GAC    4211381\n",
      "GAG    3884792\n",
      "AAG    3840330\n",
      "AAT    3533029\n",
      "ATT    3516909\n",
      "ATG    3492207\n",
      "AAC    3464665\n",
      "ATC    3245872\n",
      "GGT    3205148\n",
      "GCT    2981390\n",
      "GGC    2878725\n",
      "GCC    2842314\n",
      "GTT    2837718\n",
      "TAT    2793647\n",
      "TTT    2784500\n",
      "CTG    2766641\n",
      "TTC    2754516\n",
      "GCA    2627342\n",
      "CAG    2581736\n",
      "TAC    2462802\n",
      "GCG    2460275\n",
      "GTG    2457350\n",
      "CAA    2453443\n",
      "ACT    2379110\n",
      "ACC    2342459\n",
      "TTA    2230356\n",
      "ACA    2219802\n",
      "GTC    2206156\n",
      "GTA    2024830\n",
      "TGG    2003887\n",
      "TCT    2003845\n",
      "GGA    1952597\n",
      "CGC    1908352\n",
      "CTT    1823403\n",
      "TTG    1823265\n",
      "CTC    1802985\n",
      "CCG    1761513\n",
      "CGT    1729605\n",
      "ATA    1706275\n",
      "ACG    1657002\n",
      "CCT    1628457\n",
      "TCA    1576098\n",
      "CCA    1487845\n",
      "CAC    1469248\n",
      "AGT    1463308\n",
      "AGC    1449994\n",
      "AGA    1363662\n",
      "GGG    1349531\n",
      "TCG    1331554\n",
      "CAT    1329229\n",
      "TCC    1223679\n",
      "CTA    1151923\n",
      "CCC    1151320\n",
      "TGT     956169\n",
      "CGG     928578\n",
      "TGC     894821\n",
      "CGA     820780\n",
      "AGG     707671\n",
      "TAA     283212\n",
      "TGA     226902\n",
      "TAG      91581\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "viral_df.groupby([\"ncbi_id\"]).count()\n",
    "viral_codon_count = viral_df[viral_df.columns[8:]].sum()\n",
    "print(viral_codon_count.sort_values(ascending=False))\n",
    "# print(\"------------\")\n",
    "# print(viral_codon_count.idxmin(), viral_codon_count[viral_codon_count.idxmin()])\n",
    "# print(viral_codon_count.idxmax(), viral_codon_count[viral_codon_count.idxmax()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "codon,human,viral,alternates\n",
      "TTT,1611118,2784500,TTC\n",
      "TTC,1627492,2754516,TTT\n",
      "TTA,833315,2230356,TTG|CTT|CTC|CTA|CTG\n",
      "TTG,1278242,1823265,TTA|CTT|CTC|CTA|CTG\n",
      "TCT,1607659,2003845,TCC|TCA|TCG|AGT|AGC\n",
      "TCC,1636520,1223679,TCT|TCA|TCG|AGT|AGC\n",
      "TCA,1344428,1576098,TCT|TCC|TCG|AGT|AGC\n",
      "TCG,385999,1331554,TCT|TCC|TCA|AGT|AGC\n",
      "TAT,1138776,2793647,TAC\n",
      "TAC,1258885,2462802,TAT\n",
      "TAA,38263,283212,TAG|TGA\n",
      "TAG,30731,91581,TAA|TGA\n",
      "TGT,979680,956169,TGC\n",
      "TGC,1010784,894821,TGT\n",
      "TGA,69014,226902,TAA|TAG\n",
      "TGG,1079433,2003887,\n",
      "CTT,1339874,1823403,TTA|TTG|CTC|CTA|CTG\n",
      "CTC,1668304,1802985,TTA|TTG|CTT|CTA|CTG\n",
      "CTA,705153,1151923,TTA|TTG|CTT|CTC|CTG\n",
      "CTG,3397427,2766641,TTA|TTG|CTT|CTC|CTA\n",
      "CCT,1833292,1628457,CCC|CCA|CCG\n",
      "CCC,1805566,1151320,CCT|CCA|CCG\n",
      "CCA,1800473,1487845,CCT|CCC|CCG\n",
      "CCG,593810,1761513,CCT|CCC|CCA\n",
      "CAT,1118497,1329229,CAC\n",
      "CAC,1379660,1469248,CAT\n",
      "CAA,1332898,2453443,CAG\n",
      "CAG,3384175,2581736,CAA\n",
      "CGT,431421,1729605,CGC|CGA|CGG|AGA|AGG\n",
      "CGC,825087,1908352,CGT|CGA|CGG|AGA|AGG\n",
      "CGA,605853,820780,CGT|CGC|CGG|AGA|AGG\n",
      "CGG,1031183,928578,CGT|CGC|CGA|AGA|AGG\n",
      "ATT,1556158,3516909,ATC|ATA\n",
      "ATC,1739947,3245872,ATT|ATA\n",
      "ATA,760858,1706275,ATT|ATC\n",
      "ATG,2027598,3492207,\n",
      "ACT,1348416,2379110,ACC|ACA|ACG\n",
      "ACC,1669290,2342459,ACT|ACA|ACG\n",
      "ACA,1555392,2219802,ACT|ACC|ACG\n",
      "ACG,527773,1657002,ACT|ACC|ACA\n",
      "AAT,1747788,3533029,AAC\n",
      "AAC,1722022,3464665,AAT\n",
      "AAA,2603836,4829550,AAG\n",
      "AAG,3016615,3840330,AAA\n",
      "AGT,1340944,1463308,TCT|TCC|TCA|TCG|AGC\n",
      "AGC,1874608,1449994,TCT|TCC|TCA|TCG|AGT\n",
      "AGA,1249060,1363662,CGT|CGC|CGA|CGG|AGG\n",
      "AGG,1136933,707671,CGT|CGC|CGA|CGG|AGA\n",
      "GTT,1119006,2837718,GTC|GTA|GTG\n",
      "GTC,1260452,2206156,GTT|GTA|GTG\n",
      "GTA,725910,2024830,GTT|GTC|GTG\n",
      "GTG,2432424,2457350,GTT|GTC|GTA\n",
      "GCT,1800570,2981390,GCC|GCA|GCG\n",
      "GCC,2442252,2842314,GCT|GCA|GCG\n",
      "GCA,1619464,2627342,GCT|GCC|GCG\n",
      "GCG,560789,2460275,GCT|GCC|GCA\n",
      "GAT,2283382,4540003,GAC\n",
      "GAC,2293356,4211381,GAT\n",
      "GAA,3189350,5162667,GAG\n",
      "GAG,3771832,3884792,GAA\n",
      "GGT,1025541,3205148,GGC|GGA|GGG\n",
      "GGC,1867973,2878725,GGT|GGA|GGG\n",
      "GGA,1616254,1952597,GGT|GGC|GGG\n",
      "GGG,1442185,1349531,GGT|GGC|GGA\n"
     ]
    }
   ],
   "source": [
    "print(\"codon,human,viral,alternates\")\n",
    "for codon in CODONS_LIST:\n",
    "    print(codon, human_codon_count[codon],viral_codon_count[codon],\"|\".join(codon_alternates[codon]),sep=\",\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(Bio.Data.CodonTable.standard_dna_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
