{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Viral Immunity through Codon Optimization\n",
    "A Comparative Analysis of Genomic and Viral Data\n",
    "\n",
    "_Ishika Nelaturu_\n",
    "\n",
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Abstract\n",
    "\n",
    "REASON: The importance of my research and algorithm will help  scientists find cures & vaccines to viruses more quickly and efficiently.\n",
    "\n",
    "PROBLEM: As humans continue to develop, experiment, and grow, we as a community will face many hardships along the way. One of these hardships is the spread of new and unknown viruses. The process of finding a cure for these viruses is long and inefficient. As this process occurs, millions could die.\n",
    "\n",
    "METHODOLOGY: By using a simple algorithm, I can go through all my data to find the least used codon in human DNA & the most used codon in virus DNA. \n",
    "\n",
    "RESULT: N/A\n",
    "\n",
    "IMPLICATIONS: Using this data, scientists can find the most used codon in virus DNA & the least used codon in human DNA. If one can somehow reprogram how a ribosome processes mRNA in a way such that viruses can no longer be reproduced in a cell, this would stop the virus itself before it can do much harm. Millions of lives could be saved.\n",
    "\n",
    "The importance of my research and algorithm will help  scientists find cures & vaccines to viruses more quickly and efficiently. As humans continue to develop, experiment, and grow, we as a community will face many hardships along the way. One of these hardships is the spread of new and unknown viruses. The process of finding a cure for these viruses is long and inefficient. As this process occurs, millions could die. By using a simple algorithm, I can go through all my data to find the least used codon in human DNA & the most used codon in virus DNA.(RESULT - N/A). Using this data, scientists can find the most used codon in virus DNA & the least used codon in human DNA. If one can somehow reprogram how a ribosome processes mRNA in a way such that viruses can no longer be reproduced in a cell, this would stop the virus itself before it can do much harm. Millions of lives could be saved.\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Creation\n",
    "\n",
    "Explain how data got created\n",
    "\n",
    "[Extracting Codons data from Human Reference CDS](./NCBI-GRCh38.p14_GCF_000001405.40/process_human_codons.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: num2words in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (0.5.12)\n",
      "Requirement already satisfied: docopt>=0.6.2 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from num2words) (0.6.2)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.0.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip3 install num2words\n",
    "\n",
    "from Bio import SeqIO\n",
    "import Bio.Data.CodonTable\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from num2words import num2words\n",
    "\n",
    "import pprint\n",
    "\n",
    "pp = pprint.PrettyPrinter(indent=4)\n",
    "\n",
    "# pd.set_option('display.max_rows', None)\n",
    "# pd.options.display.max_rows = 250\n",
    "# pd.options.display.max_seq_items = 250\n",
    "\n",
    "human_df = pd.read_csv(\"./human_genes_and_proteins.tsv\", sep=\"\\t\")\n",
    "viral_df = pd.read_csv(\"./viral_genes_and_proteins.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##Codons to Amino Acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n",
      "{   'AAA': ['AAG'],\n",
      "    'AAC': ['AAT'],\n",
      "    'AAG': ['AAA'],\n",
      "    'AAT': ['AAC'],\n",
      "    'ACA': ['ACT', 'ACC', 'ACG'],\n",
      "    'ACC': ['ACT', 'ACA', 'ACG'],\n",
      "    'ACG': ['ACT', 'ACC', 'ACA'],\n",
      "    'ACT': ['ACC', 'ACA', 'ACG'],\n",
      "    'AGA': ['CGT', 'CGC', 'CGA', 'CGG', 'AGG'],\n",
      "    'AGC': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT'],\n",
      "    'AGG': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA'],\n",
      "    'AGT': ['TCT', 'TCC', 'TCA', 'TCG', 'AGC'],\n",
      "    'ATA': ['ATT', 'ATC'],\n",
      "    'ATC': ['ATT', 'ATA'],\n",
      "    'ATG': [],\n",
      "    'ATT': ['ATC', 'ATA'],\n",
      "    'CAA': ['CAG'],\n",
      "    'CAC': ['CAT'],\n",
      "    'CAG': ['CAA'],\n",
      "    'CAT': ['CAC'],\n",
      "    'CCA': ['CCT', 'CCC', 'CCG'],\n",
      "    'CCC': ['CCT', 'CCA', 'CCG'],\n",
      "    'CCG': ['CCT', 'CCC', 'CCA'],\n",
      "    'CCT': ['CCC', 'CCA', 'CCG'],\n",
      "    'CGA': ['CGT', 'CGC', 'CGG', 'AGA', 'AGG'],\n",
      "    'CGC': ['CGT', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
      "    'CGG': ['CGT', 'CGC', 'CGA', 'AGA', 'AGG'],\n",
      "    'CGT': ['CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
      "    'CTA': ['TTA', 'TTG', 'CTT', 'CTC', 'CTG'],\n",
      "    'CTC': ['TTA', 'TTG', 'CTT', 'CTA', 'CTG'],\n",
      "    'CTG': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA'],\n",
      "    'CTT': ['TTA', 'TTG', 'CTC', 'CTA', 'CTG'],\n",
      "    'GAA': ['GAG'],\n",
      "    'GAC': ['GAT'],\n",
      "    'GAG': ['GAA'],\n",
      "    'GAT': ['GAC'],\n",
      "    'GCA': ['GCT', 'GCC', 'GCG'],\n",
      "    'GCC': ['GCT', 'GCA', 'GCG'],\n",
      "    'GCG': ['GCT', 'GCC', 'GCA'],\n",
      "    'GCT': ['GCC', 'GCA', 'GCG'],\n",
      "    'GGA': ['GGT', 'GGC', 'GGG'],\n",
      "    'GGC': ['GGT', 'GGA', 'GGG'],\n",
      "    'GGG': ['GGT', 'GGC', 'GGA'],\n",
      "    'GGT': ['GGC', 'GGA', 'GGG'],\n",
      "    'GTA': ['GTT', 'GTC', 'GTG'],\n",
      "    'GTC': ['GTT', 'GTA', 'GTG'],\n",
      "    'GTG': ['GTT', 'GTC', 'GTA'],\n",
      "    'GTT': ['GTC', 'GTA', 'GTG'],\n",
      "    'TAA': ['TAG', 'TGA'],\n",
      "    'TAC': ['TAT'],\n",
      "    'TAG': ['TAA', 'TGA'],\n",
      "    'TAT': ['TAC'],\n",
      "    'TCA': ['TCT', 'TCC', 'TCG', 'AGT', 'AGC'],\n",
      "    'TCC': ['TCT', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
      "    'TCG': ['TCT', 'TCC', 'TCA', 'AGT', 'AGC'],\n",
      "    'TCT': ['TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
      "    'TGA': ['TAA', 'TAG'],\n",
      "    'TGC': ['TGT'],\n",
      "    'TGG': [],\n",
      "    'TGT': ['TGC'],\n",
      "    'TTA': ['TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
      "    'TTC': ['TTT'],\n",
      "    'TTG': ['TTA', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
      "    'TTT': ['TTC']}\n"
     ]
    }
   ],
   "source": [
    "CODONS_LIST = [\n",
    "    \"TTT\",\n",
    "    \"TTC\",\n",
    "    \"TTA\",\n",
    "    \"TTG\",\n",
    "    \"TCT\",\n",
    "    \"TCC\",\n",
    "    \"TCA\",\n",
    "    \"TCG\",\n",
    "    \"TAT\",\n",
    "    \"TAC\",\n",
    "    \"TAA\",\n",
    "    \"TAG\",\n",
    "    \"TGT\",\n",
    "    \"TGC\",\n",
    "    \"TGA\",\n",
    "    \"TGG\",\n",
    "    \"CTT\",\n",
    "    \"CTC\",\n",
    "    \"CTA\",\n",
    "    \"CTG\",\n",
    "    \"CCT\",\n",
    "    \"CCC\",\n",
    "    \"CCA\",\n",
    "    \"CCG\",\n",
    "    \"CAT\",\n",
    "    \"CAC\",\n",
    "    \"CAA\",\n",
    "    \"CAG\",\n",
    "    \"CGT\",\n",
    "    \"CGC\",\n",
    "    \"CGA\",\n",
    "    \"CGG\",\n",
    "    \"ATT\",\n",
    "    \"ATC\",\n",
    "    \"ATA\",\n",
    "    \"ATG\",\n",
    "    \"ACT\",\n",
    "    \"ACC\",\n",
    "    \"ACA\",\n",
    "    \"ACG\",\n",
    "    \"AAT\",\n",
    "    \"AAC\",\n",
    "    \"AAA\",\n",
    "    \"AAG\",\n",
    "    \"AGT\",\n",
    "    \"AGC\",\n",
    "    \"AGA\",\n",
    "    \"AGG\",\n",
    "    \"GTT\",\n",
    "    \"GTC\",\n",
    "    \"GTA\",\n",
    "    \"GTG\",\n",
    "    \"GCT\",\n",
    "    \"GCC\",\n",
    "    \"GCA\",\n",
    "    \"GCG\",\n",
    "    \"GAT\",\n",
    "    \"GAC\",\n",
    "    \"GAA\",\n",
    "    \"GAG\",\n",
    "    \"GGT\",\n",
    "    \"GGC\",\n",
    "    \"GGA\",\n",
    "    \"GGG\",\n",
    "]\n",
    "CODONS_AMINO_ACIDS_MAP = {\n",
    "    \"TTT\": \"[F] Phenylalanine\",\n",
    "    \"TTC\": \"[F] Phenylalanine\",\n",
    "    \"TTA\": \"[L] Leucine\",\n",
    "    \"TTG\": \"[L] Leucine\",\n",
    "    \"TCT\": \"[S] Serine\",\n",
    "    \"TCC\": \"[S] Serine\",\n",
    "    \"TCA\": \"[S] Serine\",\n",
    "    \"TCG\": \"[S] Serine\",\n",
    "    \"TAT\": \"[Y] Tyrosine\",\n",
    "    \"TAC\": \"[Y] Tyrosine\",\n",
    "    \"TAA\": \"[*] STOP\",\n",
    "    \"TAG\": \"[*] STOP\",\n",
    "    \"TGT\": \"[C] Cysteine\",\n",
    "    \"TGC\": \"[C] Cysteine\",\n",
    "    \"TGA\": \"[*] STOP\",\n",
    "    \"TGG\": \"[W] Tryptophan\",\n",
    "    \"CTT\": \"[L] Leucine\",\n",
    "    \"CTC\": \"[L] Leucine\",\n",
    "    \"CTA\": \"[L] Leucine\",\n",
    "    \"CTG\": \"[L] Leucine\",\n",
    "    \"CCT\": \"[P] Proline\",\n",
    "    \"CCC\": \"[P] Proline\",\n",
    "    \"CCA\": \"[P] Proline\",\n",
    "    \"CCG\": \"[P] Proline\",\n",
    "    \"CAT\": \"[H] Histidine\",\n",
    "    \"CAC\": \"[H] Histidine\",\n",
    "    \"CAA\": \"[Q] Glutamine\",\n",
    "    \"CAG\": \"[Q] Glutamine\",\n",
    "    \"CGT\": \"[R] Arginine\",\n",
    "    \"CGC\": \"[R] Arginine\",\n",
    "    \"CGA\": \"[R] Arginine\",\n",
    "    \"CGG\": \"[R] Arginine\",\n",
    "    \"ATT\": \"[I] Isoleucine\",\n",
    "    \"ATC\": \"[I] Isoleucine\",\n",
    "    \"ATA\": \"[I] Isoleucine\",\n",
    "    \"ATG\": \"[M] Methionine\",\n",
    "    \"ACT\": \"[T] Threonine\",\n",
    "    \"ACC\": \"[T] Threonine\",\n",
    "    \"ACA\": \"[T] Threonine\",\n",
    "    \"ACG\": \"[T] Threonine\",\n",
    "    \"AAT\": \"[N] Asparagine\",\n",
    "    \"AAC\": \"[N] Asparagine\",\n",
    "    \"AAA\": \"[K] Lysine\",\n",
    "    \"AAG\": \"[K] Lysine\",\n",
    "    \"AGT\": \"[S] Serine\",\n",
    "    \"AGC\": \"[S] Serine\",\n",
    "    \"AGA\": \"[R] Arginine\",\n",
    "    \"AGG\": \"[R] Arginine\",\n",
    "    \"GTT\": \"[V] Valine\",\n",
    "    \"GTC\": \"[V] Valine\",\n",
    "    \"GTA\": \"[V] Valine\",\n",
    "    \"GTG\": \"[V] Valine\",\n",
    "    \"GCT\": \"[A] Alanine\",\n",
    "    \"GCC\": \"[A] Alanine\",\n",
    "    \"GCA\": \"[A] Alanine\",\n",
    "    \"GCG\": \"[A] Alanine\",\n",
    "    \"GAT\": \"[D] Aspartic Acid\",\n",
    "    \"GAC\": \"[D] Aspartic Acid\",\n",
    "    \"GAA\": \"[E] Glutamic Acid\",\n",
    "    \"GAG\": \"[E] Glutamic Acid\",\n",
    "    \"GGT\": \"[G] Glycine\",\n",
    "    \"GGC\": \"[G] Glycine\",\n",
    "    \"GGA\": \"[G] Glycine\",\n",
    "    \"GGG\": \"[G] Glycine\",\n",
    "}\n",
    "\n",
    "AMINO_ACIDS_TO_CODONS_MAP = {}\n",
    "for codon in CODONS_AMINO_ACIDS_MAP:\n",
    "    amino_acid = CODONS_AMINO_ACIDS_MAP[codon]\n",
    "    if not amino_acid in AMINO_ACIDS_TO_CODONS_MAP:\n",
    "        AMINO_ACIDS_TO_CODONS_MAP[amino_acid] = []\n",
    "    AMINO_ACIDS_TO_CODONS_MAP[amino_acid]. append(codon)\n",
    "\n",
    "amino_acids_with_alternate_codons = {}\n",
    "for amino_acid in AMINO_ACIDS_TO_CODONS_MAP:\n",
    "    codons_list = AMINO_ACIDS_TO_CODONS_MAP[amino_acid]\n",
    "    if len(codons_list) > 0: # Change this to 1 if we can replace codon with just one alternate\n",
    "        amino_acids_with_alternate_codons[amino_acid] = codons_list\n",
    "\n",
    "print(len(amino_acids_with_alternate_codons))\n",
    "\n",
    "\n",
    "# Build codons with alternates\n",
    "codon_alternates = {}\n",
    "for clist in amino_acids_with_alternate_codons.values():\n",
    "    for codon in clist:\n",
    "        dup_list = clist[:]\n",
    "        dup_list.remove(codon)\n",
    "        codon_alternates[codon] = dup_list\n",
    "\n",
    "pp.pprint(codon_alternates)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Human Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAG      30731\n",
      "TAA      38263\n",
      "TGA      69014\n",
      "TCG     385999\n",
      "CGT     431421\n",
      "        ...   \n",
      "AAG    3016615\n",
      "GAA    3189350\n",
      "CAG    3384175\n",
      "CTG    3397427\n",
      "GAG    3771832\n",
      "Length: 64, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "human_df.groupby([\"ncbi_id\"]).count()\n",
    "human_codon_count = human_df[human_df.columns[4:]].sum()\n",
    "print(human_codon_count.sort_values())\n",
    "# print(\"------------\")\n",
    "# print(human_codon_count.idxmin(), human_codon_count[human_codon_count.idxmin()])\n",
    "# print(human_codon_count.idxmax(), human_codon_count[human_codon_count.idxmax()])"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Viral Codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAA    5162667\n",
      "AAA    4829550\n",
      "GAT    4540003\n",
      "GAC    4211381\n",
      "GAG    3884792\n",
      "        ...   \n",
      "CGA     820780\n",
      "AGG     707671\n",
      "TAA     283212\n",
      "TGA     226902\n",
      "TAG      91581\n",
      "Length: 64, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "viral_df.groupby([\"ncbi_id\"]).count()\n",
    "viral_codon_count = viral_df[viral_df.columns[8:]].sum()\n",
    "print(viral_codon_count.sort_values(ascending=False))\n",
    "# print(\"------------\")\n",
    "# print(viral_codon_count.idxmin(), viral_codon_count[viral_codon_count.idxmin()])\n",
    "# print(viral_codon_count.idxmax(), viral_codon_count[viral_codon_count.idxmax()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "codon,human,viral,alternates\n",
      "TTT,1611118,2784500,TTC\n",
      "TTC,1627492,2754516,TTT\n",
      "TTA,833315,2230356,TTG|CTT|CTC|CTA|CTG\n",
      "TTG,1278242,1823265,TTA|CTT|CTC|CTA|CTG\n",
      "TCT,1607659,2003845,TCC|TCA|TCG|AGT|AGC\n",
      "TCC,1636520,1223679,TCT|TCA|TCG|AGT|AGC\n",
      "TCA,1344428,1576098,TCT|TCC|TCG|AGT|AGC\n",
      "TCG,385999,1331554,TCT|TCC|TCA|AGT|AGC\n",
      "TAT,1138776,2793647,TAC\n",
      "TAC,1258885,2462802,TAT\n",
      "TAA,38263,283212,TAG|TGA\n",
      "TAG,30731,91581,TAA|TGA\n",
      "TGT,979680,956169,TGC\n",
      "TGC,1010784,894821,TGT\n",
      "TGA,69014,226902,TAA|TAG\n",
      "TGG,1079433,2003887,\n",
      "CTT,1339874,1823403,TTA|TTG|CTC|CTA|CTG\n",
      "CTC,1668304,1802985,TTA|TTG|CTT|CTA|CTG\n",
      "CTA,705153,1151923,TTA|TTG|CTT|CTC|CTG\n",
      "CTG,3397427,2766641,TTA|TTG|CTT|CTC|CTA\n",
      "CCT,1833292,1628457,CCC|CCA|CCG\n",
      "CCC,1805566,1151320,CCT|CCA|CCG\n",
      "CCA,1800473,1487845,CCT|CCC|CCG\n",
      "CCG,593810,1761513,CCT|CCC|CCA\n",
      "CAT,1118497,1329229,CAC\n",
      "CAC,1379660,1469248,CAT\n",
      "CAA,1332898,2453443,CAG\n",
      "CAG,3384175,2581736,CAA\n",
      "CGT,431421,1729605,CGC|CGA|CGG|AGA|AGG\n",
      "CGC,825087,1908352,CGT|CGA|CGG|AGA|AGG\n",
      "CGA,605853,820780,CGT|CGC|CGG|AGA|AGG\n",
      "CGG,1031183,928578,CGT|CGC|CGA|AGA|AGG\n",
      "ATT,1556158,3516909,ATC|ATA\n",
      "ATC,1739947,3245872,ATT|ATA\n",
      "ATA,760858,1706275,ATT|ATC\n",
      "ATG,2027598,3492207,\n",
      "ACT,1348416,2379110,ACC|ACA|ACG\n",
      "ACC,1669290,2342459,ACT|ACA|ACG\n",
      "ACA,1555392,2219802,ACT|ACC|ACG\n",
      "ACG,527773,1657002,ACT|ACC|ACA\n",
      "AAT,1747788,3533029,AAC\n",
      "AAC,1722022,3464665,AAT\n",
      "AAA,2603836,4829550,AAG\n",
      "AAG,3016615,3840330,AAA\n",
      "AGT,1340944,1463308,TCT|TCC|TCA|TCG|AGC\n",
      "AGC,1874608,1449994,TCT|TCC|TCA|TCG|AGT\n",
      "AGA,1249060,1363662,CGT|CGC|CGA|CGG|AGG\n",
      "AGG,1136933,707671,CGT|CGC|CGA|CGG|AGA\n",
      "GTT,1119006,2837718,GTC|GTA|GTG\n",
      "GTC,1260452,2206156,GTT|GTA|GTG\n",
      "GTA,725910,2024830,GTT|GTC|GTG\n",
      "GTG,2432424,2457350,GTT|GTC|GTA\n",
      "GCT,1800570,2981390,GCC|GCA|GCG\n",
      "GCC,2442252,2842314,GCT|GCA|GCG\n",
      "GCA,1619464,2627342,GCT|GCC|GCG\n",
      "GCG,560789,2460275,GCT|GCC|GCA\n",
      "GAT,2283382,4540003,GAC\n",
      "GAC,2293356,4211381,GAT\n",
      "GAA,3189350,5162667,GAG\n",
      "GAG,3771832,3884792,GAA\n",
      "GGT,1025541,3205148,GGC|GGA|GGG\n",
      "GGC,1867973,2878725,GGT|GGA|GGG\n",
      "GGA,1616254,1952597,GGT|GGC|GGG\n",
      "GGG,1442185,1349531,GGT|GGC|GGA\n"
     ]
    }
   ],
   "source": [
    "print(\"codon,human,viral,alternates\")\n",
    "for codon in CODONS_LIST:\n",
    "    print(codon, human_codon_count[codon],viral_codon_count[codon],\"|\".join(codon_alternates[codon]),sep=\",\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table 1 Standard, SGC0\n",
      "\n",
      "  |  T      |  C      |  A      |  G      |\n",
      "--+---------+---------+---------+---------+--\n",
      "T | TTT F   | TCT S   | TAT Y   | TGT C   | T\n",
      "T | TTC F   | TCC S   | TAC Y   | TGC C   | C\n",
      "T | TTA L   | TCA S   | TAA Stop| TGA Stop| A\n",
      "T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "C | CTT L   | CCT P   | CAT H   | CGT R   | T\n",
      "C | CTC L   | CCC P   | CAC H   | CGC R   | C\n",
      "C | CTA L   | CCA P   | CAA Q   | CGA R   | A\n",
      "C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "A | ATT I   | ACT T   | AAT N   | AGT S   | T\n",
      "A | ATC I   | ACC T   | AAC N   | AGC S   | C\n",
      "A | ATA I   | ACA T   | AAA K   | AGA R   | A\n",
      "A | ATG M(s)| ACG T   | AAG K   | AGG R   | G\n",
      "--+---------+---------+---------+---------+--\n",
      "G | GTT V   | GCT A   | GAT D   | GGT G   | T\n",
      "G | GTC V   | GCC A   | GAC D   | GGC G   | C\n",
      "G | GTA V   | GCA A   | GAA E   | GGA G   | A\n",
      "G | GTG V   | GCG A   | GAG E   | GGG G   | G\n",
      "--+---------+---------+---------+---------+--\n"
     ]
    }
   ],
   "source": [
    "print(Bio.Data.CodonTable.standard_dna_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_df = viral_df.groupby([\"organism\"])[viral_df.columns[8:]].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism_df.to_csv(\"./viral_codons_grouped_by_organism.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_group_df = pd.read_csv(\"./viral_codons_grouped_by_organism.csv\", sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                           organism   TTT   TTC  TTA   TTG  \\\n",
      "0                            Abaca bunchy top virus    17    24   11    12   \n",
      "1             Abalone herpesvirus Victoria/AUS/2009  1321  1162  492  1143   \n",
      "2      Abalone shriveling syndrome-associated virus   409   160  281   186   \n",
      "3                     Abelson murine leukemia virus    14    18    5    13   \n",
      "4                                      Abisko virus    75    57   66    83   \n",
      "...                                             ...   ...   ...  ...   ...   \n",
      "11529            uncultured phage_MedDCM-OCT-S42-C7   376    50  519    78   \n",
      "11530           uncultured phage_MedDCM-OCT-S45-C18   208   209  174   139   \n",
      "11531            uncultured phage_MedDCM-OCT-S45-C4   290   186  136   134   \n",
      "11532           uncultured phage_MedDCM-OCT-S46-C10   421    64  644    84   \n",
      "11533                   unidentified entomopoxvirus    12     4   10     0   \n",
      "\n",
      "       TCT  TCC  TCA  TCG  TAT  ...  GCA  GCG   GAT   GAC   GAA   GAG  GGT  \\\n",
      "0        9    2   10    7   37  ...   12   14    34     8    42    23   16   \n",
      "1      967  694  324  636  410  ...  321  837  1461  1978  2818  1857  441   \n",
      "2      227   69  202   61  230  ...  292   48   460   173   302   233  211   \n",
      "3       26   27   16    7   10  ...   17    4     9    33    18    54   10   \n",
      "4       95   40   42   31   65  ...   27   30   116    74   118    77   64   \n",
      "...    ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   ...   ...  ...   \n",
      "11529  237   17  233   37  246  ...  416   45   362   311   607   119  383   \n",
      "11530  125  118  152   95  161  ...  228  122   307   428   339   366  291   \n",
      "11531  295   88  161   52  220  ...  386   80   427   433   470   273  574   \n",
      "11532  289   17  259   24  286  ...  387   79   433   288   645   155  421   \n",
      "11533    2    0    4    0   31  ...    7    0    24     0    21     2    8   \n",
      "\n",
      "       GGC   GGA  GGG  \n",
      "0       11    21    6  \n",
      "1      478  1498  469  \n",
      "2      166   191   88  \n",
      "3       27    25   22  \n",
      "4        6    48   16  \n",
      "...    ...   ...  ...  \n",
      "11529  112   271   25  \n",
      "11530  157   195  112  \n",
      "11531  193   180   48  \n",
      "11532   89   215   39  \n",
      "11533    0    11    0  \n",
      "\n",
      "[11534 rows x 65 columns]\n"
     ]
    }
   ],
   "source": [
    "print(viral_group_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TTT</th>\n",
       "      <th>TTC</th>\n",
       "      <th>TTA</th>\n",
       "      <th>TTG</th>\n",
       "      <th>TCT</th>\n",
       "      <th>TCC</th>\n",
       "      <th>TCA</th>\n",
       "      <th>TCG</th>\n",
       "      <th>TAT</th>\n",
       "      <th>TAC</th>\n",
       "      <th>...</th>\n",
       "      <th>GCA</th>\n",
       "      <th>GCG</th>\n",
       "      <th>GAT</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>GGT</th>\n",
       "      <th>GGC</th>\n",
       "      <th>GGA</th>\n",
       "      <th>GGG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sum</th>\n",
       "      <td>2784500</td>\n",
       "      <td>2754516</td>\n",
       "      <td>2230356</td>\n",
       "      <td>1823265</td>\n",
       "      <td>2003845</td>\n",
       "      <td>1223679</td>\n",
       "      <td>1576098</td>\n",
       "      <td>1331554</td>\n",
       "      <td>2793647</td>\n",
       "      <td>2462802</td>\n",
       "      <td>...</td>\n",
       "      <td>2627342</td>\n",
       "      <td>2460275</td>\n",
       "      <td>4540003</td>\n",
       "      <td>4211381</td>\n",
       "      <td>5162667</td>\n",
       "      <td>3884792</td>\n",
       "      <td>3205148</td>\n",
       "      <td>2878725</td>\n",
       "      <td>1952597</td>\n",
       "      <td>1349531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>241</td>\n",
       "      <td>238</td>\n",
       "      <td>193</td>\n",
       "      <td>158</td>\n",
       "      <td>173</td>\n",
       "      <td>106</td>\n",
       "      <td>136</td>\n",
       "      <td>115</td>\n",
       "      <td>242</td>\n",
       "      <td>213</td>\n",
       "      <td>...</td>\n",
       "      <td>227</td>\n",
       "      <td>213</td>\n",
       "      <td>393</td>\n",
       "      <td>365</td>\n",
       "      <td>447</td>\n",
       "      <td>336</td>\n",
       "      <td>277</td>\n",
       "      <td>249</td>\n",
       "      <td>169</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>558</td>\n",
       "      <td>364</td>\n",
       "      <td>534</td>\n",
       "      <td>310</td>\n",
       "      <td>355</td>\n",
       "      <td>191</td>\n",
       "      <td>267</td>\n",
       "      <td>317</td>\n",
       "      <td>585</td>\n",
       "      <td>343</td>\n",
       "      <td>...</td>\n",
       "      <td>400</td>\n",
       "      <td>597</td>\n",
       "      <td>792</td>\n",
       "      <td>775</td>\n",
       "      <td>846</td>\n",
       "      <td>546</td>\n",
       "      <td>500</td>\n",
       "      <td>631</td>\n",
       "      <td>322</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>median</th>\n",
       "      <td>77</td>\n",
       "      <td>79</td>\n",
       "      <td>36</td>\n",
       "      <td>75</td>\n",
       "      <td>59</td>\n",
       "      <td>37</td>\n",
       "      <td>53</td>\n",
       "      <td>26</td>\n",
       "      <td>67</td>\n",
       "      <td>65</td>\n",
       "      <td>...</td>\n",
       "      <td>85</td>\n",
       "      <td>37</td>\n",
       "      <td>133</td>\n",
       "      <td>92</td>\n",
       "      <td>137</td>\n",
       "      <td>108</td>\n",
       "      <td>72</td>\n",
       "      <td>49</td>\n",
       "      <td>73</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>...</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "      <td>11534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            TTT      TTC      TTA      TTG      TCT      TCC      TCA  \\\n",
       "sum     2784500  2754516  2230356  1823265  2003845  1223679  1576098   \n",
       "mean        241      238      193      158      173      106      136   \n",
       "std         558      364      534      310      355      191      267   \n",
       "median       77       79       36       75       59       37       53   \n",
       "count     11534    11534    11534    11534    11534    11534    11534   \n",
       "\n",
       "            TCG      TAT      TAC  ...      GCA      GCG      GAT      GAC  \\\n",
       "sum     1331554  2793647  2462802  ...  2627342  2460275  4540003  4211381   \n",
       "mean        115      242      213  ...      227      213      393      365   \n",
       "std         317      585      343  ...      400      597      792      775   \n",
       "median       26       67       65  ...       85       37      133       92   \n",
       "count     11534    11534    11534  ...    11534    11534    11534    11534   \n",
       "\n",
       "            GAA      GAG      GGT      GGC      GGA      GGG  \n",
       "sum     5162667  3884792  3205148  2878725  1952597  1349531  \n",
       "mean        447      336      277      249      169      117  \n",
       "std         846      546      500      631      322      196  \n",
       "median      137      108       72       49       73       48  \n",
       "count     11534    11534    11534    11534    11534    11534  \n",
       "\n",
       "[5 rows x 64 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_group_df[viral_group_df.columns[1:]].agg(['sum', 'mean', 'std', 'median', 'count']).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_group_df[viral_group_df.columns[1:]].agg(['std', 'sum', 'mean', 'median', 'count']).astype(int).transpose().to_csv(\"./viral_sum_mean_std.csv\")\n",
    "\n",
    "# sort_values(by = ['Rank', 'Age'], ascending = [True, False], na_position = 'first')\n",
    "\n",
    "# viral_group_df[viral_group_df.columns[1:]].agg(['std', 'sum', 'mean', 'median', 'count']).astype(int).transpose().sort_values(by = ['std', 'sum', 'mean', 'median'], ascending = [True, False, True, True])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TTT    241.416681\n",
       "TTC    238.817063\n",
       "TTA    193.372291\n",
       "TTG    158.077423\n",
       "TCT    173.733744\n",
       "          ...    \n",
       "GAG    336.812207\n",
       "GGT    277.886943\n",
       "GGC    249.586007\n",
       "GGA    169.290532\n",
       "GGG    117.004595\n",
       "Length: 64, dtype: float64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_group_df[viral_group_df.columns[1:]].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TTT    558.379736\n",
       "TTC    364.658301\n",
       "TTA    534.926693\n",
       "TTG    310.692623\n",
       "TCT    355.032034\n",
       "          ...    \n",
       "GAG    546.648716\n",
       "GGT    500.718915\n",
       "GGC    631.522295\n",
       "GGA    322.295012\n",
       "GGG    196.262288\n",
       "Length: 64, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_group_df[viral_group_df.columns[1:]].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "viral_group_df[viral_group_df.columns[1:]].agg(['std', 'sum', 'mean', 'median', 'count', 'skew']).astype(int).transpose().to_csv(\"./viral_data_with_skew.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMKYPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITKEKKKELQDCKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGEVRNDDVDQSLIIAARNIVRRATVSADPLASLLEMCHSTQIGGIRMVDILRQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTSGSSVKREEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGKDEQSIAEAIIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSKMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTEKLTITYSSSMMWEINGPESVLVNTYQWIIRNWENVKIQWSQDPTMLYNKMEFEPFQSLVPKAARGQYSGFVRVLFQQMRDVLGTFDTVQIIKLLPFAAAPPKQSRMQFSSLTVNVRGSGMRIVVRGNSPVFNYNKATKRLTVLGKDAGALMEDPDEGTAGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*\n",
      "MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPALRMKWMMAMKYPITADKRIMEMIPERNEQ*QTLWSKTNDAGSDRVMVSPLAVTWWNRNGPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQDVIMEVVFPNEVGARILTSESQLTITKEKKKELQDCKIAPLMVAYMLERELVRKTRFLPVAGGTSSVYIEVLHLTQGTCWEQMYTPGGEVRNDDVDQSLIIAARNIVRRATVSADPLASLLEMCHSTQIGGIRMVDILRQNPTEEQAVDICKAAM*LRISSSFSFG*FTFKRTSGSSVKREEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGKDEQSIAEAIIVAMVFSQEDCMIKAVR*DLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWGIEPIDNVMGMIGILPDMTPSTEMSLRGVRVSKMGVDEYSSTERVVVSIDRFLRVRDQRGNVLLSPEEVSETQGTEKLTITYSSSMMWEIN*PESVLVNTYQWIIRNWENVKIQWSQDPTMLYNKMEFEPFQSLVPKAARGQYSGFVRVLFQQMRDVLGTFDTVQIIKLLPFAAAPPKQSRMQFSSLTVNVR*SGMRIVVRGNSPVFNYNKATKRLTVLGKDA*ALMEDPDEGTAGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRDSSILTDSQTATKRIRMAIN*\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "influenza_A_cds_PB2_YP_009118470_1 = \"ATGGAAAGAATAAAAGAACTAAGAGATTTGATGTCACAGTCTCGCACTCGCGAGATACTGACAAAAACAACTGTGGACCATATGGCCATAATCAAGAAATATACATCAGGAAGACAGGAGAAGAATCCTGCCCTTAGGATGAAGTGGATGATGGCAATGAAATATCCAATTACGGCAGACAAAAGGATAATGGAGATGATCCCGGAAAGAAATGAGCAAGGTCAGACCCTTTGGAGCAAGACAAATGATGCTGGATCAGACAGAGTGATGGTGTCACCTCTGGCTGTGACGTGGTGGAACAGAAATGGACCAACGACAAGCACAGTCCATTATCCAAAGGTCTATAAAACCTATTTTGAAAAGGTCGAAAGGCTAAAACATGGAACCTTCGGCCCCGTTCACTTCAGAAACCAGGTTAAAATACGCCGCAGGGTCGACATAAACCCGGGCCATGCAGATCTTAGTGCTAAAGAAGCACAGGATGTCATCATGGAGGTCGTATTCCCAAACGAAGTTGGAGCCAGAATATTGACATCAGAGTCACAGTTAACGATTACCAAGGAAAAGAAGAAGGAGCTTCAGGACTGCAAAATTGCCCCTTTAATGGTGGCTTACATGTTGGAGAGAGAACTGGTTCGCAAAACAAGATTCCTACCAGTAGCTGGAGGGACAAGCAGCGTGTATATCGAGGTGTTGCATTTGACCCAAGGGACCTGCTGGGAGCAAATGTACACACCGGGAGGGGAAGTGAGAAATGATGATGTTGATCAGAGTTTAATTATTGCTGCTAGAAATATTGTTAGAAGAGCAACAGTATCAGCAGACCCGTTGGCTTCGCTTTTGGAGATGTGCCATAGTACACAGATTGGCGGGATTAGGATGGTTGACATCCTTAGACAAAACCCAACAGAAGAACAGGCTGTGGATATATGTAAGGCAGCAATGGGTCTAAGGATCAGTTCATCCTTCAGCTTTGGAGGTTTCACTTTCAAAAGGACAAGTGGGTCATCTGTCAAAAGGGAAGAAGAAGTGCTCACAGGCAACCTCCAAACATTGAAAATAAGAGTACATGAAGGATATGAGGAATTCACAATGGTCGGGCGAAGAGCAACAGCCATTCTAAGGAAAGCAACCAGAAGACTGATCCAACTGATAGTGAGTGGGAAAGACGAGCAATCAATCGCCGAGGCAATCATAGTGGCAATGGTATTCTCACAAGAGGATTGTATGATAAAGGCAGTGAGAGGTGATTTGAACTTTGTCAACAGAGCAAACCAGCGGCTAAATCCCATGCATCAACTCCTGAGGCATTTCCAAAAGGATGCAAAGGTCCTGTTTCAAAACTGGGGAATTGAACCCATTGACAATGTAATGGGGATGATCGGAATATTGCCTGACATGACCCCCAGCACAGAGATGTCATTGAGAGGAGTGAGAGTTAGTAAAATGGGAGTAGATGAATATTCCAGCACTGAGAGAGTGGTCGTGAGTATTGATCGTTTCTTGAGGGTCCGAGACCAGAGGGGAAACGTACTCCTGTCTCCTGAAGAGGTTAGTGAAACACAGGGAACAGAAAAGCTGACTATAACATATTCATCGTCCATGATGTGGGAGATCAATGGTCCGGAATCAGTGCTAGTTAACACATATCAATGGATCATTAGAAATTGGGAAAATGTAAAGATTCAATGGTCCCAAGATCCTACAATGCTATACAATAAGATGGAATTTGAACCCTTTCAATCCCTAGTGCCTAAAGCTGCCAGGGGCCAATATAGTGGGTTCGTGAGGGTTCTATTCCAACAGATGCGTGACGTACTGGGAACATTTGACACTGTCCAAATAATAAAGCTATTACCATTTGCAGCAGCCCCGCCGAAGCAGAGTAGGATGCAGTTCTCTTCTCTAACTGTGAATGTGAGGGGTTCCGGAATGAGAATAGTTGTGAGAGGCAATTCTCCTGTGTTCAACTACAACAAGGCAACCAAGAGGCTTACGGTGCTTGGAAAGGATGCAGGTGCATTGATGGAAGACCCCGATGAGGGAACAGCAGGAGTGGAATCTGCGGTATTGAGGGGATTTCTGATTCTGGGCAAAGAAGACAAAAGATATGGGCCAGCATTGAGCATCAACGAATTGAGCAATCTTGCGAAAGGAGAGAAGGCTAATGTGTTGATAGGGCAAGGAGACGTTGTGTTGGTGATGAAACGGAAACGGGACTCTAGCATACTTACTGACAGTCAGACAGCGACCAAAAGGATTCGGATGGCCATCAATTAA\"\n",
    "\n",
    "print(Seq(influenza_A_cds_PB2_YP_009118470_1).translate())\n",
    "\n",
    "# Original from genbank file\n",
    "\"\"\"\n",
    "/translation=\"MERIKELRDLMSQSRTREILTKTTVDHMAIIKKYTSGRQEKNPA\n",
    "LRMKWMMAMKYPITADKRIMEMIPERNEQGQTLWSKTNDAGSDRVMVSPLAVTWWNRN\n",
    "GPTTSTVHYPKVYKTYFEKVERLKHGTFGPVHFRNQVKIRRRVDINPGHADLSAKEAQ\n",
    "DVIMEVVFPNEVGARILTSESQLTITKEKKKELQDCKIAPLMVAYMLERELVRKTRFL\n",
    "PVAGGTSSVYIEVLHLTQGTCWEQMYTPGGEVRNDDVDQSLIIAARNIVRRATVSADP\n",
    "LASLLEMCHSTQIGGIRMVDILRQNPTEEQAVDICKAAMGLRISSSFSFGGFTFKRTS\n",
    "GSSVKREEEVLTGNLQTLKIRVHEGYEEFTMVGRRATAILRKATRRLIQLIVSGKDEQ\n",
    "SIAEAIIVAMVFSQEDCMIKAVRGDLNFVNRANQRLNPMHQLLRHFQKDAKVLFQNWG\n",
    "IEPIDNVMGMIGILPDMTPSTEMSLRGVRVSKMGVDEYSSTERVVVSIDRFLRVRDQR\n",
    "GNVLLSPEEVSETQGTEKLTITYSSSMMWEINGPESVLVNTYQWIIRNWENVKIQWSQ\n",
    "DPTMLYNKMEFEPFQSLVPKAARGQYSGFVRVLFQQMRDVLGTFDTVQIIKLLPFAAA\n",
    "PPKQSRMQFSSLTVNVRGSGMRIVVRGNSPVFNYNKATKRLTVLGKDAGALMEDPDEG\n",
    "TAGVESAVLRGFLILGKEDKRYGPALSINELSNLAKGEKANVLIGQGDVVLVMKRKRD\n",
    "SSILTDSQTATKRIRMAIN\"\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "codons_list = [influenza_A_cds_PB2_YP_009118470_1[i:i+3] for i in range(0, len(influenza_A_cds_PB2_YP_009118470_1), 3)]\n",
    "# print(codons_list)\n",
    "# replace GGT with TAG (one of stop codon)\n",
    "modified_codons_list = [(\"TAG\" if codon == \"GGT\" else codon) for codon in codons_list]\n",
    "# print(modified_codons_list)\n",
    "\n",
    "print(Seq(\"\".join(modified_codons_list)).translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
